src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -29,113 +31,152 @@ public static class WebUtility
 
         #region HtmlEncode / HtmlDecode methods
 
-        private static readonly char[] s_htmlEntityEndingChars = new char[] { ';', '&' };
-
         public static string HtmlEncode(string value)
         {
             if (string.IsNullOrEmpty(value))
             {
                 return value;
             }
 
-            // Don't create StringBuilder if we don't have anything to encode
-            int index = IndexOfHtmlEncodingChars(value, 0);
+            ReadOnlySpan<char> valueSpan = value.AsSpan();
+
+            // Don't create ValueStringBuilder if we don't have anything to encode
+            int index = IndexOfHtmlEncodingChars(valueSpan);
             if (index == -1)
             {
                 return value;
             }
 
-            StringBuilder sb = StringBuilderCache.Acquire(value.Length);
-            HtmlEncode(value, index, sb);
-            return StringBuilderCache.GetStringAndRelease(sb);
+            char[] rawBuffer = null;
+            Span<char> buffer = value.Length < 80 ?
+                stackalloc char[256] :
+                (rawBuffer = ArrayPool<char>.Shared.Rent(value.Length + 200));
+            var sb = new ValueStringBuilder(buffer);
+
+            sb.Append(valueSpan.Slice(0, index));
+            HtmlEncode(valueSpan.Slice(index), ref sb);
+            string result = sb.ToString();
+
+            if (rawBuffer != null)
+            {
+                ArrayPool<char>.Shared.Return(rawBuffer);
+            }
```

Comment:
Why are we doing it this way rather than the previous approach of passing the initial capacity into VSB and having it do the ArrayPool access?  With this approach, we're going to hold onto this ArrayPool buffer no matter how many times the VSB grows, whereas if the VSB did it, it would return the original buffer to the pool when it first grew out of it.